File: DataStructures/Stacks/leetcode/20. Valid Parentheses/ValidParentheses.js
// 承知しました。
// 以下は、指定された括弧の有効性判定問題を\*\*JavaScript（Node.js 18.16.1 / LeetCode対応）\*\*でクラスを使わず関数形式で解いた実装です。

// ---

// ## ✅ **問題再掲：**

// **有効な括弧列**とは：

// * `()[]{}` など、**開き括弧と閉じ括弧が正しく対応しているもの**
// * 開いた順に対応する閉じ括弧で閉じられているもの

// ---

// ## 🔧 **実装（JavaScript）**

// ```javascript
/**
 * 与えられた文字列sが有効な括弧列か判定する関数
 * 
 * @param {string} s - 入力文字列（括弧のみ： '()[]{}'）
 * @returns {boolean} - 有効な括弧列ならtrue、不正ならfalse
 * 
 * 制約: 1 <= s.length <= 10^4
 * 
 * 時間計算量: O(n) （nは文字列の長さ）
 * 空間計算量: O(n) （スタックを使用するため）
 */
function isValidJs(s) {
    const stack = [];
    
    const bracketMap = {
        ')': '(',
        '}': '{',
        ']': '['
    };

    for (let char of s) {
        if (char === '(' || char === '{' || char === '[') {
            stack.push(char);
        } else {
            if (stack.length === 0) {
                return false;
            }
            const top = stack.pop();
            if (top !== bracketMap[char]) {
                return false;
            }
        }
    }

    return stack.length === 0;
}
// ```

// ---

// ## ✅ **実行例**

// ```javascript
// console.log(isValid("()"));       // true
// console.log(isValid("()[]{}"));   // true
// console.log(isValid("(]"));       // false
// console.log(isValid("([])"));     // true
// console.log(isValid("([)]"));     // false
// ```

// ---

// ## 🔍 **処理時間・メモリ解析**

// | 項目        | 計算量                 |
// | --------- | ------------------- |
// | **時間計算量** | O(n)（各文字を一度だけ処理）    |
// | **空間計算量** | O(n)（スタック使用時の最大サイズ） |

// ---

// ## ⚙️ **LeetCode対応ポイント**

// * **Node.js 18.16.1対応**
// * クラスを使わず関数形式
// * `console.log` によるテスト実行も可能
// * `stack`を使うことで括弧の入れ子構造を適切に管理

// ---

// もし、追加で**図解やアルゴリズム解説、ブラウザ動作版**が欲しければ教えてください！

File: DataStructures/Stacks/leetcode/20. Valid Parentheses/ValidParentheses.py
# 以下は、**LeetCode用のPython実装（CPython 3.11.4）**で、
# 指定の括弧の妥当性判定を**クラス `Solution` 内メソッドで実装**したものです。

# ## ✅ 実装コード（Python）

# ```python
from typing import List

class Solution:
    def isValid(self, s: str) -> bool:
        """
        与えられた文字列sが有効な括弧列か判定する関数

        パラメーター:
        s: str - 入力文字列（括弧のみ： '()[]{}'）

        返却値:
        bool - 有効な括弧列ならTrue、不正ならFalse

        制約:
        1 <= len(s) <= 10^4

        時間計算量: O(n) （nは文字列長）
        空間計算量: O(n) （スタック使用）
        """
        stack: List[str] = []
        
        bracket_map = {
            ')': '(',
            '}': '{',
            ']': '['
        }

        for char in s:
            if char in '({[':
                stack.append(char)
            else:
                if not stack:
                    return False
                top = stack.pop()
                if bracket_map[char] != top:
                    return False

        return not stack
# ```

# ---

# ## ✅ **使用例（テストコード）**

# ```python
# # テスト実行例
# solution = Solution()
# print(solution.isValid("()"))        # True
# print(solution.isValid("()[]{}"))    # True
# print(solution.isValid("(]"))        # False
# print(solution.isValid("([])"))      # True
# print(solution.isValid("([)]"))      # False
# ```

# ---

# ## 🔍 **解析**

# | 項目        | 計算量                          |
# | --------- | ---------------------------- |
# | **時間計算量** | O(n) （各文字を一度だけ処理）            |
# | **空間計算量** | O(n) （最悪ケースでスタックに全ての開き括弧が入る） |

# ---

# ## 🔧 **特徴**

# * **LeetCode用構造** → `class Solution` に定義
# * **型ヒント付き**（`List[str]`、`-> bool`）
# * **最適化済み** → 余計な処理なし、シンプルかつ効率的

# ---

# もし、追加で**スタック動作の図解、再帰解法や例外ケースの説明**が必要であればお伝えください！
